### DISCUSSION TITLE
Using own data set for transferring in shadow-Tor instead of  generated traffics with TGen

### DISCUSSION THREAD

**Discussion by Reyhanea** [shadow/shadow#672] [General]
I would really appreciate your help and kindness in a problem based on your experience with shadow. I need to test the efficiency of the Tor network in transmitting a special data set with shadow. In fact, instead of using generated data of TGen simulated traffic generator for transmitting data, I want to transfer my own data set through the Tor network and evaluate its performance. I kindly ask you if it is possible to guide me on how I could do it with shadow?
I wonder if you could kindly help me.

**Comment by robgjansen** [comment]
If you want to replace TGen, then you would need your own program to transfer the data. Usually you would have a client side program that connects to a Tor client on the Tor SOCKS port, and you would have a server side program that the client connects to through Tor (i.e., the exit relay connects to the server and sends the client-generated requests).

Do you have your own protocol? Or are you trying to download standard webpages through shadow-tor?

If you want to create non-standard traffic streams, it would not be that difficult to write your own client/server programs in C and compile them as Shadow plugins. There are several examples of doing this here:
https://github.com/shadow/shadow-plugin-extras

**Comment by Reyhanea** [comment]
-In my scenario, each client downloads a standard web page as well as should send a file to the server.  I need to this transferred file be a file in my own data set.
In the provided examples, the code at the client side only tries to send some random information (not file, image and so). I need to know is it possible to have an example in shadow-Tor that the client sends a file or image to a simulated server?

- I should make my own server and then try this scenario?

**Comment by robgjansen** [comment]
There are some existing plug-ins that you might want to have a look at:
https://github.com/shadow/shadow-plugin-extras

Specifically, the `web` set of plugins may be useful for you:
https://github.com/shadow/shadow-plugin-extras/tree/master/web

The only problem is that this plugin worked for Shadow v1.9.2, and now we are at Shadow 1.13.0 and I don't think it will work anymore without a little bit of fixing. Still, it might give you a good starting point.

Otherwise, you can always start fresh and write your own client/server - that way you can keep it minimal and make sure it only does what you need.

**Comment by robgjansen** [comment]
I recently moved the tgen plugin outside of Shadow, so this provides a good example of how to compile existing code as an executable that can then be run in shadow:
https://github.com/shadow/tgen

I'm marking this as solved, let us know if you have other questions.
